### PR TITLE
base: Add atomic CAS macros/functions

### PR DESCRIPTION
**Pull request by nicowilliams**

### CODE REVIEW

File: lib/base/heimbase-atomics.h [hunk 1]
```diff
@@ -151,6 +164,10 @@ heim_base_exchange_64(uint64_t *p, uint64_t newval)
     return val;
 }
 
+#define heim_base_cas_pointer(t,e,d) 	compare_and_swaplp((atomic_l)(t), (e), (d))
+#define heim_base_cas_32(t,e,d)		compare_and_swap  ((atomic_l)(t), (e), (d))
```
Comment (lhoward) [inline lib/base/heimbase-atomics.h:168]: Believe this should be `atomic_p` for 32-bit (see `heim_base_atomic_dec_32()`)

Comment (lhoward) [reply lib/base/heimbase-atomics.h:168]: https://www.ibm.com/docs/en/aix/7.2?topic=c-compare-swap-kernel-services

File: lib/base/heimbase-atomics.h [hunk 2]
```diff
@@ -164,6 +181,10 @@ heim_base_exchange_64(uint64_t *p, uint64_t newval)
 #define heim_base_exchange_32(t,v)	((ULONG)InterlockedExchange((LONG volatile *)(t), (LONG)(v)))
 #define heim_base_exchange_64(t,v)	((ULONG64)InterlockedExchange64((LONG64 violatile *)(t), (LONG64)(v)))
 
+#define heim_base_cas_pointer(t,e,d) 	InterlockedCompareExchangePointer((atomic_l)(t), (d), (e))
+#define heim_base_cas_32(t,e,d)		InterlockedCompareExchange  ((atomic_l)(t), (d), (e))
+#define heim_base_cas_64(t,e,d)		InterlockedCompareExchange64((atomic_l)(t), (d), (e))
+
```
Comment (lhoward) [inline lib/base/heimbase-atomics.h:187]: Windows will want `PVOID volatile *`, `LONG volatile *`, `LONG64 volatile *` respectively. If they even need casts?

Comment (nicowilliams) [reply lib/base/heimbase-atomics.h:187]: Oops, good catch!

A test that uses these (for what?) would expose this.

**Comment by nicowilliams** [comment]
I've reworked this a fair bit and tested `<stdatomic.h>` and GCC atomic builtins.  There is now also a test so we'll know that this works on Windows.  The test is not very good in that it doesn't race threads, but it should elicit compilation errors/warnings if `<heimdal-atomics.h>` has errors.

**Comment by nicowilliams** [comment]
I stopped the UBSAN clang build to save runner time allocation.

**Pull request closed without merging**